mongo/integration/server_selection_prose_test.go

```diff
@@ -21,9 +23,80 @@ import (
 	"go.mongodb.org/mongo-driver/mongo/options"
 )
 
+// ? do we want to extend saturation set as an mtest.T method?
+
+// saturationSet is used to maintain information about events with specific host+pool combinations.
+type saturationSet map[string]bool
+
+func (set saturationSet) has(host string, connectionID uint64) bool {
+	return set[host+strconv.FormatInt(int64(connectionID), 10)]
+}
+
+func (set saturationSet) add(host string, connectionID uint64) {
+	set[host+strconv.FormatInt(int64(connectionID), 10)] = true
+}
+
+func (set saturationSet) isUnsaturated(mt *mtest.T, maxPoolSize uint64) bool {
+	hosts := options.Client().ApplyURI(mtest.ClusterURI()).Hosts
+	return uint64(len(set)) < maxPoolSize*uint64(len(hosts))
+}
+
+// awaitSaturation uses CMAP events to ensure that the client's connection pools for N-mongoses have been saturated.
+// The qualification for a host to be "saturated" is for that host to have the maximum number of connections allowed by
+// the test, in this case `maxPoolSize`.
+func awaitSaturation(mt *mtest.T, monitor *monitor.TestPoolMonitor, maxPoolSize uint64) {
```

Comment:
What happens if the saturation condition never occurs? We may need a way to time-out `awaitSaturation` to allow graceful test failure instead of timing out the entire test run. Adding that timeout logic will also allow us to print a more useful test failure message that describes why the saturation condition wasn't met (e.g. what hosts were missing connections).

Reply:
I added an update that I _think_ will satisfy this.